services/worker/src/worker/job_runners/split/duckdb_index.py

```diff
@@ -0,0 +1,250 @@
+        raise ParquetResponseEmptyError("No parquet files found.")
+
+    # get the features
+    features = content_parquet_and_info["dataset_info"].get("features", [])
+
+    # look for string columns using the first rows
+    string_columns = [
+        column
+        for column, feature in features.items()
+        if "dtype" in feature
+        and "_type" in feature
+        and feature["dtype"] == STRING_FEATURE_DTYPE
+        and feature["_type"] == VALUE_FEATURE_TYPE
+    ]
+    if not string_columns:
+        raise NoIndexableColumnsError("No string columns available to index.")
+
+    # configure duckdb extensions
+    duckdb.execute(INSTALL_EXTENSION_COMMAND.format(extension="httpfs"))
+    duckdb.execute(LOAD_EXTENSION_COMMAND.format(extension="httpfs"))
+    duckdb.execute(INSTALL_EXTENSION_COMMAND.format(extension="fts"))
+    duckdb.execute(LOAD_EXTENSION_COMMAND.format(extension="fts"))
+
+    # index all columns
+    db_location = f"{duckdb_index_file_directory}/{DUCKDB_DEFAULT_INDEX_FILENAME}"
+    con = duckdb.connect(str(db_location))
+    con.sql(CREATE_SEQUENCE_COMMAND)
+    con.sql(f"{CREATE_TABLE_COMMAND} read_parquet({parquet_urls});")
+
+    # TODO: by default, 'porter' stemmer is being used, use a specific one by dataset language in the future
+    # see https://duckdb.org/docs/extensions/full_text_search.html for more details about 'stemmer' parameter
+    con.sql(CREATE_INDEX_COMMAND)
+
+    # create the target revision if it does not exist yet (clone from initial commit to avoid cloning all repo's files)
+    hf_api = HfApi(endpoint=hf_endpoint, token=hf_token)
+    committer_hf_api = HfApi(endpoint=hf_endpoint, token=committer_hf_token)
+    index_file_location = f"{config}/{split}/{DUCKDB_DEFAULT_INDEX_FILENAME}"
+    try:
+        refs = hf_api.list_repo_refs(repo_id=dataset, repo_type=DATASET_TYPE)
+        if all(ref.ref != target_revision for ref in refs.converts):
+            initial_commit = hf_api.list_repo_commits(repo_id=dataset, repo_type=DATASET_TYPE)[-1].commit_id
+            committer_hf_api.create_branch(
+                repo_id=dataset, branch=target_revision, repo_type=DATASET_TYPE, revision=initial_commit, exist_ok=True
+            )
+    except RepositoryNotFoundError as err:
+        raise DatasetNotFoundError("The dataset does not exist on the Hub.") from err
+
+    target_dataset_info = hf_api.dataset_info(repo_id=dataset, revision=target_revision, files_metadata=False)
+    all_repo_files: Set[str] = {f.rfilename for f in target_dataset_info.siblings}
+    delete_operations: List[CommitOperation] = []
+    if index_file_location in all_repo_files:
+        delete_operations.append(CommitOperationDelete(path_in_repo=index_file_location))
+
+    # send the files to the target revision
+    add_operations: List[CommitOperation] = [
+        CommitOperationAdd(path_in_repo=index_file_location, path_or_fileobj=db_location)
+    ]
+
+    committer_hf_api.create_commit(
```

Comment:
Here we'll need to use the new locking mechanism to avoid concurrent commits like in `https://github.com/huggingface/datasets-server/pull/1355`

Maybe let's agree on using something like
```python
lock_key = json.dumps({"dataset": dataset, "resource": "parquet conversion branch"})
with lock(key=lock_key, job_id=job_id):
    committer_hf_api.create_commit(...)
```

I can do the change in my PR to use this

Reply:
Thanks, I applied it as in parquet_and_info.